### PR TITLE
fix: step size is made dynamic

### PR DESCRIPTION
**Pull request by palashgdev**
Close https://github.com/SigNoz/signoz/issues/2812
Close https://github.com/SigNoz/signoz/issues/2986

### CODE REVIEW

File: frontend/src/lib/getStep.ts [hunk 1]
```diff
@@ -40,7 +40,10 @@ const getStep = ({ start, end, inputFormat = 'ms' }: GetStepInput): number => {
 	const endDate = dayjs(convertToMs(Number(end), inputFormat));
 	const diffSec = Math.abs(endDate.diff(startDate, 's'));
 
-	return Math.max(Math.floor(diffSec / MaxDataPoints), DefaultStepSize);
+	return (
+		Math.max(Math.floor(diffSec / MaxDataPoints), DefaultStepSize) ||
+		DefaultStepSize
```
Comment (makeavish) [inline frontend/src/lib/getStep.ts:45]: Why are we using `|| DefaultStepSize` here?

Comment (palashgdev) [reply frontend/src/lib/getStep.ts:45]: it's fallback to default step size

File: frontend/src/lib/getStep.ts [hunk 2]
```diff
@@ -40,7 +40,10 @@ const getStep = ({ start, end, inputFormat = 'ms' }: GetStepInput): number => {
 	const endDate = dayjs(convertToMs(Number(end), inputFormat));
 	const diffSec = Math.abs(endDate.diff(startDate, 's'));
 
-	return Math.max(Math.floor(diffSec / MaxDataPoints), DefaultStepSize);
+	return (
```
Comment (srikanthccv) [inline frontend/src/lib/getStep.ts:43]: Add tests for this.

File: frontend/src/container/GridGraphLayout/Graph/utils.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import { GlobalReducer } from 'types/reducer/globalTime';
+
+export const useStep = (query: Widgets['query']): Widgets['query'] => {
+	const { maxTime, minTime } = useSelector<AppState, GlobalReducer>(
+		(state) => state.globalTime,
+	);
+
+	return {
+		...query,
+		builder: {
```
[the remark this answers is not in the document]
Comment (palashgdev) [reply frontend/src/container/GridGraphLayout/Graph/utils.tsx:14]: in metrics application we don't promql query so no need and this lib only take care of composite query

File: frontend/src/lib/getStep.ts (hunk 2 above)
Comment (palashgdev) [reply frontend/src/lib/getStep.ts:43]: sure

File: frontend/src/container/GridGraphLayout/Graph/utils.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (srikanthccv) [reply frontend/src/container/GridGraphLayout/Graph/utils.tsx:14]: I am not talking about the metrics application I am referring to a case where the user is using the dashboards with the promQL tab.

[the remark this answers is not in the document]
Comment (palashgdev) [reply frontend/src/container/GridGraphLayout/Graph/utils.tsx:14]: this only take care of the composite Query `frontend/src/store/actions/dashboard/getQueryResults.ts` take care of the promql step

**Comment by palashgdev** [comment]
> Will test the functionality locally.

sure :-D

**Comment by srikanthccv** [comment]
- breaks when you create a new dashboard and click on add a panel.
- it doesn't work in panel edit mode. the `stepInterval` value is still sent as 30 for the builder query

**Comment by palashgdev** [comment]
> * breaks when you create a new dashboard and click on add a panel.
> * it doesn't work in panel edit mode. the `stepInterval` value is still sent as 30 for the builder query

happened after merge from develop please re check

**Comment by srikanthccv** [comment]
- what's reflected in url, builder aggregate every is different from what's sent in API call
Screenshot 2023-06-17 at 6 05 58 AM

**Comment by palashgdev** [comment]
> * what's reflected in url, builder aggregate every is different from what's sent in API call

my bad ... please re check

**Review by makeavish — APPROVED** [review]
Address this issue separately: https://github.com/SigNoz/signoz/issues/2931
LGTM, otherwise

**Comment by srikanthccv** [comment]
I will test this again.

**Comment by srikanthccv** [comment]
- If the user provides a step interval larger than the calculated interval, we should honour it. We only need to override it when it's lower than the calculated.
- Following up on the last comment https://github.com/SigNoz/signoz/pull/2903#issuecomment-1595519844, I think it opens a can of worms  https://www.loom.com/share/8224bb4134ab44b9823968e06e77fa5f

**Comment by palashgdev** [comment]
> * If the user provides a step interval larger than the calculated interval, we should honour it. We only need to override it when it's lower than the calculated.
> * Following up on the last comment [fix: step size is made dynamic #2903 (comment)](https://github.com/SigNoz/signoz/pull/2903#issuecomment-1595519844), I think it opens a can of worms  https://www.loom.com/share/8224bb4134ab44b9823968e06e77fa5f

glad you brought though this one as these new changes will make pr open again when things are clear

**Comment by srikanthccv** [comment]
https://www.loom.com/share/85baf5c78366479aaf2ba405171d300a

**Comment by palashgdev** [comment]
> https://www.loom.com/share/85baf5c78366479aaf2ba405171d300a

updated @srikanthccv

**Comment by srikanthccv** [comment]
I think this was the only issue I could find, let me check again.

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**